### PR TITLE
Add support for Grist embeds.

### PR DESCRIPTION
**Pull request by prio**
As per https://github.com/outline/outline/discussions/3851.

Supports the SaaS Grist product and self-hosted versions.

### CODE REVIEW

File: shared/editor/embeds/Grist.tsx [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import * as React from "react";
+import Frame from "../components/Frame";
+import { EmbedProps as Props } from ".";
+
+const URL_REGEX = new RegExp(
+  "^https?://.*/p/([0-9][0-9]*)\\?embed=true$" // Ensure its a page URL
```
Comment (apoorv-mishra) [inline shared/editor/embeds/Grist.tsx:6]: The regex seems to be too weak. I'm assuming it's kept this way in order to support the possibility of Grist being self hosted. But, did you also consider the possibility of this being interfering/colliding with other supported embeds? What do you think @tommoor ?

Comment (tommoor) [reply shared/editor/embeds/Grist.tsx:6]: Agreed that it's too loose, suggest we change to only support saas-hosted to make this mergable and then revisit how to support self-hosted.

A pattern is needed for this (and some other integrations) where there is a section under Settings -> Integrations that you can enter your self-hosted url which will then be used to calculate the match here.

Comment (prio) [reply shared/editor/embeds/Grist.tsx:6]: Yes, that is why. I forced the /p/1?embed=true at the end but thats as much as I could do. Any suggestions besides change it to only work on the SaaS product?

Comment (apoorv-mishra) [reply shared/editor/embeds/Grist.tsx]: > Any suggestions besides changing it to only work on the SaaS product?

I second @tommoor 's suggestion on this. Honestly, a weaker regex might find its place in a specific, closed setting as a quick hack but pushing it upstream is not a good idea owing to unforeseen, negative consequences.

Comment (prio) [reply shared/editor/embeds/Grist.tsx]: Code updated. It would be great to support self-hosted versions at some point. (that is one of the main reasons I support Outline over competing products).

Comment (tommoor) [reply shared/editor/embeds/Grist.tsx]: Yep, as mentioned above we need to prioritize the pattern as it'll benefit multiple integrations including draw.io/diagrams.net as well.

File: shared/editor/embeds/Grist.tsx [hunk 2]
```diff
@@ -0,0 +1,17 @@
+import * as React from "react";
+import Frame from "../components/Frame";
+import { EmbedProps as Props } from ".";
+
+const URL_REGEX = new RegExp(
+  "^https?://([a-z.-]+\\.)?getgrist\\.com/(.*)/p/([0-9][0-9]*)\\?embed=true$"
```
[the remark this answers is not in the document]
Comment (prio) [reply shared/editor/embeds/Grist.tsx]: Change applied

File: shared/editor/embeds/Grist.tsx [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+const URL_REGEX = new RegExp(
+  "^https?://([a-z.-]+\\.)?getgrist\\.com/(.*)/p/([0-9]+)\\?embed=true$"
+);
+
+export default class Grist extends React.Component<Props> {
+  static ENABLED = [URL_REGEX];
+
+  render() {
+    return (
+      <Frame {...this.props} src={this.props.attrs.href} title="Grist Embed" />
```
Comment (apoorv-mishra) [inline shared/editor/embeds/Grist.tsx:14]: Screenshot from 2022-08-06 00-31-46

Just realized that one small improvement you could do here is to have a footer as shown in the sample screenshot. Use [this](https://github.com/outline/outline/blob/74d9409cc32c5923fe72e98eb961d1d56304519e/shared/editor/embeds/GoogleDocs.tsx#L13-L27) as reference.

File: shared/editor/embeds/Grist.tsx [hunk 4]
```diff
@@ -0,0 +1,17 @@
+import * as React from "react";
+import Frame from "../components/Frame";
+import { EmbedProps as Props } from ".";
+
+const URL_REGEX = new RegExp(
+  "^https?://([a-z.-]+\\.)?getgrist\\.com/(.*)/p/([0-9]+)\\?embed=true$"
```
Comment (apoorv-mishra) [inline shared/editor/embeds/Grist.tsx:6]: Also, are you sure that all the pasted urls would always have `embed=true` at the end? If not, then you might want to loosen this a bit and later replace the trailing part of the url with `embed=true` as demonstrated [here](https://github.com/outline/outline/blob/74d9409cc32c5923fe72e98eb961d1d56304519e/shared/editor/embeds/GoogleDocs.tsx#L15)

File: shared/editor/embeds/index.tsx [hunk 5]
```diff
@@ -233,6 +234,13 @@ const embeds: EmbedDescriptor[] = [
     component: GoogleDataStudio,
     matcher: matcher(GoogleDataStudio),
   },
+  {
+    title: "Grist",
+    keywords: "project",
```
Comment (apoorv-mishra) [inline shared/editor/embeds/index.tsx:239]: Nitpick :)
```suggestion
    keywords: "spreadsheet",
```

File: shared/editor/embeds/Grist.tsx [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+);
+
+export default class Grist extends React.Component<Props> {
```
Comment (apoorv-mishra) [inline shared/editor/embeds/Grist.tsx:9]: Apologies for intermittent and disruptive reviews. It's so because I'm currently on a beginner's track and just getting familiar with the practices followed within the project.

One of the practices is that the newly developed components should preferably be functional instead of class-based. Therefor, you might want to change this. You may use [`Gliffy.tsx`](https://github.com/outline/outline/blob/74d9409cc32c5923fe72e98eb961d1d56304519e/shared/editor/embeds/Gliffy.tsx) as reference.

File: shared/editor/embeds/Grist.tsx (hunk 4 above)
Comment (prio) [reply shared/editor/embeds/Grist.tsx:6]: Yes, they must have embed=true at the end

File: shared/editor/embeds/Grist.tsx (hunk 6 above)
Comment (prio) [reply shared/editor/embeds/Grist.tsx:9]: Ok, Ill try and revisit it tonight or tomorrow

File: shared/editor/embeds/Grist.tsx (hunk 4 above)
Comment (tommoor) [reply shared/editor/embeds/Grist.tsx:6]: Not sure how Grist works but generally the idea with these embeds is that we take a regular url and transform it into the embedded version on the users behalf

File: shared/editor/embeds/Grist.tsx (hunk 6 above)
Comment (prio) [reply shared/editor/embeds/Grist.tsx]: Ok, new version committed now

File: shared/editor/embeds/Grist.tsx (hunk 4 above)
Comment (prio) [reply shared/editor/embeds/Grist.tsx]: Ok, that would be useful actually, let me look at that

File: shared/editor/embeds/Grist.tsx [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+      }
+      title="Grist Spreadsheet"
+      border
+    />
+  );
+}
+
+Grist.ENABLED = [
+  new RegExp(
+    "^(https?://[a-z.-]+\\.?getgrist\\.com/.*/p/[0-9]+)(\\?embed=true)?$"
```
Comment (apoorv-mishra) [inline shared/editor/embeds/Grist.tsx:29]: Couple of concerns here...

1. It matches `https://fakegetgrist.com/doc/fake/p/1`, we surely don't want this :)
2. Maybe the part `/p/[0-9]+` is too strict. The [`src`](https://github.com/gristlabs/grist-core/blob/9434b8e4549713212c1ad5b0796a2e9db2511cd6/app/common/gristUrls.ts#L174) seems to suggest that `p` is followed by `docPage`, which in itself could be a [number or a string](https://github.com/gristlabs/grist-core/blob/9434b8e4549713212c1ad5b0796a2e9db2511cd6/app/common/gristUrls.ts#L82). Moreover, seems like `p/<docPage` is optional.

I'd suggest
```suggestion
    "^https?://([a-z.-]+\\.)?getgrist\\.com/doc/(.+)$"
```

File: shared/editor/embeds/Grist.tsx [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+import Image from "../components/Image";
+import { EmbedProps as Props } from ".";
+
+function Grist(props: Props) {
+  const { matches } = props.attrs;
+  const sheetURL = matches[1];
+  return (
+    <Frame
+      {...props}
+      src={`${sheetURL}?embed=true`}
```
Comment (apoorv-mishra) [inline shared/editor/embeds/Grist.tsx:12]: Needs adjustment after [this](https://github.com/outline/outline/pull/3914/commits/09fba4b4aecc0b964efd637796c0504756a36ea5) :)
```suggestion
      src={props.attrs.href.replace(/(\?embed=true)?$/, "?embed=true")}
```

**Review by apoorv-mishra — COMMENTED** [review]
Could you resolve conflicts please?

**Comment by prio** [comment]
> Could you resolve conflicts please?

@apoorv-mishra This is done

**Comment by prio** [comment]
Hi @apoorv-mishra 

This request has been open 8 days now. Is it going to be merged?

File: shared/editor/embeds/Grist.test.tsx [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+  const match = Grist.ENABLED[0];
+
+  test("to be enabled on share link", () => {
+    expect(
+      "https://templates.getgrist.com/doc/afterschool-program/p/2?embed=true".match(
+        match
+      )
+    ).toBeTruthy();
+    expect(
+      "https://docs.getgrist.com/sg5V93LuAije/Untitled-document/p/22?embed=true".match(
```
Comment (apoorv-mishra) [inline shared/editor/embeds/Grist.test.tsx:13]: Did you test this locally?

Why aren't the `ci` tests running against this PR? Any idea @tommoor ?

**Review by apoorv-mishra — COMMENTED** [review]
Sorry, got caught up with something. Left a few comments :point_down:

File: shared/editor/embeds/Grist.test.tsx (hunk 9 above)
Comment (tommoor) [reply shared/editor/embeds/Grist.test.tsx:13]: Not sure, building of forks is enabled in CircleCI. Could be that @prio needs to follow the project there or something.

Comment (prio) [reply shared/editor/embeds/Grist.test.tsx:13]: > Did you test this locally?

Yes, tested it locally

Comment (prio) [reply shared/editor/embeds/Grist.test.tsx:13]: > Not sure, building of forks is enabled in CircleCI. Could be that @prio needs to follow the project there or something.

Ok, I am now following the project on circleci

File: shared/editor/embeds/Grist.tsx [hunk 10]
```diff
@@ -22,10 +22,6 @@ function Grist(props: Props) {
   );
 }
 
-Grist.ENABLED = [
-  new RegExp(
-    "^https?://([a-z.-]+\\.)?getgrist\\.com/doc/(.+)$"
-  ),
-];
+Grist.ENABLED = [new RegExp("^https?://([a-z.-]+\\.)?getgrist\\.com/(.+)$")];
```
Comment (apoorv-mishra) [inline shared/editor/embeds/Grist.tsx:25]: Sorry, why was this changed?

```suggestion
Grist.ENABLED = [new RegExp("^https?://([a-z.-]+\\.)?getgrist\\.com/doc/(.+)$")];
```

File: shared/editor/embeds/Grist.test.tsx (hunk 9 above)
Comment (apoorv-mishra) [reply shared/editor/embeds/Grist.test.tsx:13]: > Yes, tested it locally

I meant unit([`jest`](https://github.com/outline/outline/blob/0c51bfb8998a08e5565a9e9cd5c4f19a3fcd0df6/package.json#L27) ones), not manual testing. Anyways, good that CI started running tests.

File: shared/editor/embeds/Grist.tsx (hunk 10 above)
Comment (prio) [reply shared/editor/embeds/Grist.tsx:25]: https://docs.getgrist.com/sg5V93LuAije/Untitled-document/p/22?embed=true

Is a valid URL. "doc" is not required.

Comment (apoorv-mishra) [reply shared/editor/embeds/Grist.tsx:25]: Cool then. My bad!

**Review by tommoor — APPROVED** [review]

**Comment by tommoor** [comment]
> It would be great to support self-hosted versions at some point. (that is one of the main reasons I support Outline over competing products).

Just a quick followup – @apoorv-mishra is actively working on this now, so it's going to be a fast-follow rather than a "someday" feature 💪
